verilog/tools/kythe/kythe_facts_extractor.cc

```diff
@@ -0,0 +1,153 @@
+
+namespace verilog {
+namespace kythe {
+
+void ExtractKytheFacts(IndexingFactNode& node) {
+  KytheFactsExtractor kythe_extractor(node);
+  kythe_extractor.Visit(node);
+}
+
+void KytheFactsExtractor::Visit(IndexingFactNode& node) {
+  const auto tag =
+      static_cast<IndexingFactType>(node.Value().GetIndexingFactType());
+  switch (tag) {
+    case IndexingFactType::kFile: {
+      ExtractFileFact(node);
+      break;
+    }
+    case IndexingFactType::kModule: {
+      ExtractModuleFact(node);
+      break;
+    }
+    case IndexingFactType::kModuleInstance: {
+      ExtractModuleInstanceFact(node);
+      break;
+    }
+    case IndexingFactType::kModulePortInput:
+      break;
+    case IndexingFactType::kModulePortOutput:
+      break;
+  }
+
+  const AutoPop p(&facts_tree_context_, &node);
+  for (auto& child : node.Children()) {
+    Visit(child);
+  }
+}
+
+void KytheFactsExtractor::ExtractFileFact(IndexingFactNode& node) {
+  VName file_vname(file_path_, "", "");
+
+  PrintFact(file_vname, kFactNodeKind, kNodeFile);
+  PrintFact(file_vname, kFactText, node.Value().Anchors()[1].Value());
+}
+
+void KytheFactsExtractor::ExtractModuleFact(IndexingFactNode& node) {
+  VName module_vname(file_path_,
+                     CreateModuleSignature(node.Value().Anchors()[0].Value()));
+
+  PrintFact(module_vname, kFactNodeKind, kNodeRecord);
+  PrintFact(module_vname, kFactSubkind, kSubkindModule);
+  PrintFact(module_vname, kFactComplete, kCompleteDefinition);
+
+  VName module_name_anchor =
+      PrintAnchorVname(node.Value().Anchors()[0], file_path_);
+
+  PrintEdge(module_name_anchor, kEdgeDefinesBinding, module_vname);
+
+  if (node.Value().Anchors().size() > 1) {
+    VName module_end_label_anchor =
+        PrintAnchorVname(node.Value().Anchors()[1], file_path_);
+    PrintEdge(module_end_label_anchor, kEdgeRef, module_vname);
+  }
+}
+
+void KytheFactsExtractor::ExtractModuleInstanceFact(IndexingFactNode& node) {
+  VName module_instance_vname(
+      file_path_,
+      CreateModuleInstantiationSignature(node.Value().Anchors()[1].Value()));
+  VName module_instance_anchor =
+      PrintAnchorVname(node.Value().Anchors()[1], file_path_);
+  VName module_type_vname(
+      file_path_, CreateModuleSignature(node.Value().Anchors()[0].Value()));
+  VName module_type_anchor =
+      PrintAnchorVname(node.Value().Anchors()[0], file_path_);
+
+  PrintFact(module_instance_vname, kFactNodeKind, kNodeVariable);
+  PrintFact(module_instance_vname, kFactComplete, kCompleteDefinition);
+
+  PrintEdge(module_type_anchor, kEdgeRef, module_type_vname);
+  PrintEdge(module_instance_vname, kEdgeTyped, module_type_vname);
+  PrintEdge(module_instance_anchor, kEdgeDefinesBinding, module_instance_vname);
+}
+
+std::string PrintVName(const VName& vname) {
+  return absl::Substitute(
+      R"({"signature": "$0","path": "$1","language": "$2","root": "$3","corpus": "$4"})",
+      vname.signature, vname.path, vname.language, vname.root, vname.corpus);
```

Comment:
Would it make sense to move this to `std::ostream& operator<<(std::ostream&, const VName&)`?